### PR TITLE
New Algorithm: Overlapping Normalized Mutual Information

### PR DESCRIPTION
**Pull request by jonasspinner**
This implements the Overlapping Normalized Mutual Information from [NMI13]. I'm happy to receive some feedback.

There are some points which might need to be changed.

1. Documentation. Most of the algorithm is implemented in separate functions. Some of the doc comments might be unnecessary. The inclusion of LaTeX formulas could be too much and the formatting could also be wrong.
2. Naming conventions. The public methods conform to camelCase, but the hidden functions and variable names are in snake_case. The only attribute is called `m_normalization`. The string parameters for normalization are also in snake_case.
3. Normalization (`normalize`). The current implementation allows the specification of the method used to normalize the mutual information. The authors use "max". My impression is that it is the most useful choice. It could be better to remove the ability to choose the normalization method.
4. Numerical error handling (`clamp_below`, `clamp_above`). To make sure that the value actually is in [0, 1], it is clamped. Discrepancies less than `Aux::NumericTools::acceptableError` are silently corrected, otherwise they are corrected and an error is logged with `ERRORF`. I have only seen errors within the acceptable error range, but this might not be true for all inputs.
5. Only looking at intersections (`conditional_entropy`). I(X;Y) is defined as a sum of minima of H\*(X_i|Y_j) over all cluster pairs (X_i, Y_j) between the covers. The [reference implementation](https://github.com/aaronmcdaid/Overlapping-NMI) from [NMI13] and my code only calculate H\*(X_i|Y_j) if X_i and Y_j intersect. In the case that there exists a X_i that does not intersect with any Y_j, the entropy of X_i is used. This can lead to different results. A fix for that case should be possible as described in the comments of `conditional_entropy`, but I'm not sure if it is worth the additional complexity. A simple fix would be to iterate over all pairs of clusters, but this would make the algorithm very inefficient for covers with a lot of clusters, O(|X||Y|) instead of O(number of intersecting pairs of clusters).

[NMI13]
McDaid, Aaron F., Derek Greene, and Neil Hurley. "Normalized Mutual Information to Evaluate Overlapping
Community Finding Algorithms." ArXiv:1110.2515 [Physics], August 2, 2013. http://arxiv.org/abs/1110.2515.

### CODE REVIEW

**Comment by jonasspinner** [comment]
Regarding only looking at intersecting pairs of clusters. An example is (X_i, Y_j) where X_i and X_j do not intersect and have 
0 to 100 nodes. The error occurs at around (|X_i|, |Y_j|) = (80, 3) and (3, 80).

entropies
selection

The maximum error is 9.3076 at (3, 87).
errors

**Comment by jonasspinner** [comment]
The non-succesful CI builds seem to fail at installing prequisites. The commits pass the CI builds with my fork. Not sure how to handle that.

**Comment by fabratu** [comment]
The CI-errors happen occasionally due to updates in the underlying Linux-images (syncing of package-repositories). After a restart the tests now are successful.

**Comment by avdgrinten** [comment]
This addition fits well into NetworKit. Thank you for the PR.

The code looks good in general but let's wait for a proper code review to see if any changes need to be done.

File: networkit/cpp/community/OverlappingNMIDistance.cpp
Comment (angriman) [inline networkit/cpp/community/OverlappingNMIDistance.cpp:30]: This might be useful also in other parts of NetworKit, it's probably a good idea to move it to the `auxiliary` module.

File: networkit/cpp/community/OverlappingNMIDistance.cpp
Comment (angriman) [inline networkit/cpp/community/OverlappingNMIDistance.cpp:40]: I think that wrapping the result into a `struct` is cleaner than using a tuple.

File: networkit/cpp/community/OverlappingNMIDistance.cpp
Comment (angriman) [inline networkit/cpp/community/OverlappingNMIDistance.cpp:312]: I think it's a better idea to implement these utility methods as private static members of `OverlappingNMIDistance`.
Also, please use `camelBack`.

File: networkit/cpp/community/OverlappingNMIDistance.cpp
Comment (jonasspinner) [reply networkit/cpp/community/OverlappingNMIDistance.cpp]: That seems like a good idea. I added `Aux::hashCombine` and `Aux::PairHash` in a `HashUtils.hpp` (e9590cd).

File: networkit/cpp/community/OverlappingNMIDistance.cpp
Comment (jonasspinner) [reply networkit/cpp/community/OverlappingNMIDistance.cpp]: Done (b073403)

File: networkit/cpp/community/OverlappingNMIDistance.cpp
Comment (jonasspinner) [reply networkit/cpp/community/OverlappingNMIDistance.cpp]: Done (4685253 and 18f2d4d)

**Review by angriman — APPROVED** [review]
Thank you, the PR looks good to me. However, I am not very familiar with this topic so I encourage other NetworKit developers/maintainers to review it.

**Review by angriman — APPROVED** [review]

File: include/networkit/auxiliary/HashUtils.hpp
Comment (jonasspinner) [inline include/networkit/auxiliary/HashUtils.hpp:13]: This might need to be changed or include the [Boost Software License](https://www.boost.org/users/license.html).

Comment (hmeyerhenke) [reply include/networkit/auxiliary/HashUtils.hpp:13]: I'd rather avoid to include another license. It simply complicates matters and I wonder if there is no simpler other way.

Comment (jonasspinner) [reply include/networkit/auxiliary/HashUtils.hpp:13]: A simple combination of two hash values which is better than just an xor would be:
```
constexpr auto halfWidth = std::numeric_limits<size_t>::digits / 2;
auto a = std::hash<A>()(pair.first);
auto b = std::hash<B>()(pair.second);
return a ^ (b << halfWidth);
```

Comment (avdgrinten) [reply include/networkit/auxiliary/HashUtils.hpp:13]: I think it is enough to add something like `The Boost license applies to this function.` Boost's license is non-viral and compatible with the MIT license.

Comment (hmeyerhenke) [reply include/networkit/auxiliary/HashUtils.hpp:13]: If that's the case, then fine.

Comment (jonasspinner) [reply include/networkit/auxiliary/HashUtils.hpp]: Thank you for the input. I added a license notice to the function.

**Review by michitux — APPROVED** [review]
I think this looks good. I have only one minor note which is how to represent the enum in Python. Currently, it is not exported and instead a string is used. It is possible to export the enum to Python, but then again NetworKit uses string parameters in a lot of other places, too, so I think this can stay as it is. Concerning the normalization I wouldn't mind keeping the choice, in particular as `NMIDistance` currently does not use the `max` normalization. It would make sense to also implement at least the `max` normalization for `NMIDistance`.

Just as a side a note, @jonasspinner implemented this as part of the practical part of my lecture on graph partitioning and graph clustering.

**Comment by jonasspinner** [comment]
That's good to hear.

By coincidence, I was just looking into how to replace it with an enum. There seem to be different variants used in Networkit like `MaxentStress` with  [cdef enum](https://github.com/networkit/networkit/blob/e6a9ffe92b926d984d5c0961c396f87463e64dde/networkit/viz.pyx#L98) and [integer member constants](https://github.com/networkit/networkit/blob/e6a9ffe92b926d984d5c0961c396f87463e64dde/networkit/viz.pyx#L160), `MultipleEdgesHandling` with a [class on module level](https://github.com/networkit/networkit/blob/a9bcf06d32ffe40fcb58878af0c62b4cf92ae966/networkit/graphio.pyx#L73), or as you have mentioned with just using strings.

I also encountered problems with using a `enum class` in the C++ class in combination with the `cdef enum` in cython, but it should be [possible](https://stackoverflow.com/questions/31001918/wrap-enum-class-for-cython).

So I would say that either an python enum `OverlappingNMIDistanceNormalization` in `community` or integer constants like `OverlappingNMIDistance.MAX` seem to be possible for me. But I am not really familiar with cython, so there might be a better way.

**Comment by michitux** [comment]
I do not think integer constants are a good solution as inconsistencies between C++ and Python could easily be introduced. I implemented something similar to the solution you mentioned, see [here](https://github.com/michitux/networkit/blob/f2610311f7d948ba1889e35af5cbe570b2a62468/networkit/community.pyx#L1109-L1137) for an example (that is not yet part of NetworKit). Using strings seems to be a common pattern in Python as far as I am aware, so I do not want to say that this needs to be changed. A consistent approach in NetworKit would make sense, though.

**Comment by jonasspinner** [comment]
I rebased and squashed the commits, decided for using the enum in cython and removed `nogil` annotations.

**Comment by michitux** [comment]
Why did you remove the `nogil` annotations? Using `nogil` allows running code in parallel from threads in Python and is in general a good idea for methods whose running time can be long.

**Comment by jonasspinner** [comment]
I misunderstood the effects/the need for `nogil`. I thought that it is only needed for parallelism in C++. I will fix that shortly. Thanks for catching that.

**Review by michitux — APPROVED** [review]
Thanks, I think this is good now. I still think there should be a standard how to expose enums/choices to Python but for now I would suggest to merge this.

**Pull request merged**
